ramips: add support for Kroks KNdRt31R16 and KNdRt31R19 routers

bam80 opened the pull request:
Adding two Kroks devices:
- ramips: add support for KROKS Rt-Cse SIM Injector DS

    Aka Kroks Rt-Cse5 UW DRSIM (KNdRt31R16), ID 1958:
    https://kroks.ru/search/?text=1958
    See Kroks OpenWrt fork for support of other models:
    https://github.com/kroks-free/openwrt
    
    Device specs:
    - CPU: MediaTek MT7628AN
    - Flash: 16MB SPI NOR
    - RAM: 64MB
    - Bootloader: U-Boot
    - Ethernet: 5x 10/100 Mbps
    - 2.4 GHz: b/g/n SoC
    - USB: 1x
    - SIM-reader: 2x (driven by a dedicated chip with it's own firmware)
    - Buttons: reset
    - LEDs: 1x Power, 1x Wi-Fi, 12x others (SIM status, Internet, etc.)
    
    Flashing:
    - sysupgrade image via stock firmware WEB interface, IP: 192.168.1.254
    - U-Boot launches a WEB server if Reset button is held during power up,
      IP: 192.168.1.1
    
    MAC addresses as verified by OEM firmware:
    vendor   OpenWrt   source
    LAN      eth0      factory 0x4 (label)
    2g       wlan0     label
    
    Signed-off-by: Andrey Butirsky <butirsky@gmail.com>


- ramips: add support for Kroks Rt-Pot mXw DS RSIM router
    
    Aka "Kroks KNdRt31R19".
    Ported from v19.07.8 of OpenWrt fork:
    see https://github.com/kroks-free/openwrt
    for support of other models.
    
    Device specs:
    - CPU: MediaTek MT7628AN
    - Flash: 16MB SPI NOR
    - RAM: 64MB
    - Bootloader: U-Boot
    - Ethernet: 1x 10/100 Mbps
    - 2.4 GHz: b/g/n SoC
    - mPCIe: 1x (usually equipped with an LTE modem by vendor)
    - Buttons: reset
    - LEDs: 1x Modem, 1x Injector, 1x Wi-Fi, 1x Status
    
    Flashing:
    - sysupgrade image via stock firmware WEB interface.
    - U-Boot launches a WEB server if Reset button is held during power up.
    Server IP: 192.168.1.1
    
    SIM card switching:
    The device supports up to 4 SIM cards - 2 locally on board and 2 on
    remote SIM-injector.
    By default, 1-st local SIM is active.
    To switch to e.g. 1-st remote SIM:
    echo 0 > /sys/class/gpio/modem1power/value
    echo 0 > /sys/class/gpio/modem1sim1/value
    echo 1 > /sys/class/gpio/modem1rsim1/value
    echo 1 > /sys/class/gpio/modem1power/value
    
    MAC addresses as verified by OEM firmware:
    vendor   OpenWrt   source
    LAN      eth0      factory 0x4 (label)
    2g       wlan0     label
    
    Signed-off-by: Kroks <dev@kroks.ru>
    [butirsky@gmail.com: port to master; drop dts-v1]
    Signed-off-by: Andrey Butirsky <butirsky@gmail.com>

adschm commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -31,6 +31,7 @@ ampedwireless,ally-r1900k)
 	;;
 buffalo,wsr-1166dhp|\
 buffalo,wsr-600dhp|\
+kroks,kndrt31r*|\
```
No wildcards here, please.
https://openwrt.org/docs/guide-developer/device-support-policies#no_wildcards

bam80 replied:
done

adschm commented on target/linux/ramips/dts/flash_kroks_16m.dtsi:
```diff
@@ -0,0 +1,12 @@
+&partitions {
```
This PR only adds one device, so please just create one DTS file with all the nodes put together. We may split up stuff when we actually have more devices.

bam80 replied:
done

adschm commented on target/linux/ramips/dts/flash_kroks_16m.dtsi:
```diff
@@ -0,0 +1,12 @@
+
+	partition@f00000 {
+		label = "memory";
+		reg = <0xf00000 0x100000>;
```
read-only?

bam80 replied:
done

adschm commented on target/linux/ramips/dts/mt7628an_kroks.dtsi:
```diff
@@ -0,0 +1,42 @@
+#include "mt7628an.dtsi"
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
```
We use nvmem-cells now. With this, MAC address will be random during test. Did you test with master at all?

bam80 replied:
No, will do ASAP when I get the access to the device.

adschm commented on target/linux/ramips/dts/mt7628an_kroks.dtsi:
```diff
@@ -0,0 +1,42 @@
+};
+
+&wmac {
+	status = "okay";
```
Add an empty line after status.

bam80 replied:
done

adschm commented on target/linux/ramips/dts/mt7628an_kroks.dtsi:
```diff
@@ -0,0 +1,42 @@
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+		m25p,fast-read;
```
Fast-read is typically only required for 50 MHz+. Does this device run stable with 50 MHz BTW?

bam80 replied:
It looks like it's needed to work around the problem with race condition in SPI driver on MT7628, after firmware update.
I don't know the details nor references to actual problem, though.

bam80 replied:
Might be related to https://github.com/openwrt/openwrt/pull/1743 but I'm completely unaware. @kroks-free 
Is it any harm to keep it with `fast-read`?

adschm commented on target/linux/ramips/dts/mt7628an_kroks.dtsi:
```diff
@@ -0,0 +1,42 @@
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+			};
```
3x read-only

bam80 replied:
done

adschm commented on target/linux/ramips/dts/mt7628an_kroks_kndrt31r19.dts:
```diff
@@ -0,0 +1,75 @@
+#include "mt7628an_kroks_1eth.dtsi"
```
Missing SPDX license identifier.

bam80 replied:
done

adschm commented on target/linux/ramips/dts/mt7628an_kroks_kndrt31r19.dts:
```diff
@@ -0,0 +1,75 @@
+
+/ {
+	compatible = "kroks,kndrt31r19", "mediatek,mt7628an-soc";
+	model = "Kroks KNdRt31R19";
```
This is completely different from the title in the PR?

bam80 replied:
Well, it seems like the device is actually has 2 names - the internal one here and some marketing name in the title.
Both names are used in original commit: 
`DEVICE_TITLE := Kroks Rt-Pot mXw DS RSIM`:
https://github.com/kroks-free/openwrt/commit/ba19321f71c680cbbec2da5d60ee5f9947a33eca#diff-bb176229946c2cb6b2c4899e91408e84b49fe2518003b294e577f8605145c220R187
and
`model = "Kroks KNdRt31R19";` here:
https://github.com/kroks-free/openwrt/commit/ba19321f71c680cbbec2da5d60ee5f9947a33eca#diff-c30ff05179792fdffab9c2bdef800c31b9ad56ddf9366da24f29e3af86d4d045R5

I don't know how to handle that properly, but I think keeping them both might help to identify the device.

adschm commented on target/linux/ramips/dts/mt7628an_kroks_kndrt31r19.dts:
```diff
@@ -0,0 +1,75 @@
+		compatible = "gpio-leds";
+
+		led_status: status {
+			label = "kndrt31r19:led:status";
```
model prefix should be dropped, just "led:status" etc.

bam80 replied:
done (it was suggested I should write "done")

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -169,6 +169,29 @@ define Device/Default
+	kmod-usb-wdm \
+	usb-modeswitch \
+	zram-swap
+endef
```
Do we really need all of these for boot?
Packages should be added at the device node.

bam80 replied:
The device supposed to be connected to inet via modem inserted in it's mPCIe slot, only.
So I guess idea behind that is to have it capable for initial inet connection with stock firmware.
We can skip `zram-swap` but I'm not sure about the others modules.

The device comes equipped with Quectel EP-06 modem or some Huaway modem, but optionally might come without the modem at all.

bam80 replied:
Still I wonder what the other similar "modem boards" can suggest here?

For the reference, here is the same device with slightly different "marketing" name:
https://kroksw.com/shop/3g-modems-and-routers/routers-with-sim-injector/router-kroks-rt-ubx-rsim-eq-ep-with-m-pci-modem-lte-cat6-quectel-ep06-e-and-sim-injector/

bam80 replied:
I reduced the packages list to a bare minimum to get LTE connection and thus make the device with no other WAN ports usable out of the box.

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -240,6 +240,15 @@ define Device/jotale_js76x8-32m
 TARGET_DEVICES += jotale_js76x8-32m
 
+define Device/kroks_kndrt31r19
+  IMAGE_SIZE := 14m
```
Please use 15040k instead.

bam80 replied:
I'm not sure if we really need such a big image as I removed `memory` partition in the end (1MB, extended `firmware` to the same amount), so the tail now is `rootfs_data`:
```
root@OpenWrt:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00030000 00010000 "u-boot"
mtd1: 00010000 00010000 "u-boot-env"
mtd2: 00010000 00010000 "factory"
mtd3: 00fb0000 00010000 "firmware"
mtd4: 001eb18d 00010000 "kernel"
mtd5: 00dc4e73 00010000 "rootfs"
mtd6: 00a30000 00010000 "rootfs_data"

bam80 replied:
> Please use 15040k instead.

I've removed `memory` partition (it only served vendor's needs and unused normally) and thus extended `firmware` to 1MiB.
As seen above, the new `firmware` size is 00fb0000 which is 16064k, this value is used as IMAGE_SIZE.
Please reopen if my calculations are wrong.

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -240,6 +240,15 @@ define Device/jotale_js76x8-32m
+define Device/kroks_kndrt31r19
+  IMAGE_SIZE := 14m
+  DEVICE_VENDOR := Kroks
+  DEVICE_MODEL := Rt-Pot mXw DS RSIM
```
Again, Model does not match the device node name.

bam80 replied:
This is what is exposed to Web UI I suppose, so I just kept what was in original commit here:
https://github.com/kroks-free/openwrt/commit/ba19321f71c680cbbec2da5d60ee5f9947a33eca#diff-bb176229946c2cb6b2c4899e91408e84b49fe2518003b294e577f8605145c220R187

Please reopen if we need the match anyway.

bam80 replied:
Actually, I didn't find any presence of this in the UI but it listed in `menuconfig` at least.

hauke replied:
You can add something like this:
````
  DEVICE_ALT0_VENDOR := ASUS
  DEVICE_ALT0_MODEL := RT-N12+
  DEVICE_ALT0_VARIANT := B1
````
To give it two names.

bam80 replied:
Done, thank you

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -240,6 +240,15 @@ define Device/jotale_js76x8-32m
+  DEVICE_VENDOR := Kroks
+  DEVICE_MODEL := Rt-Pot mXw DS RSIM
+  $(Device/Kroks-mpci)
+  SUPPORTED_DEVICES += kndrt31r19
```
What's this line required for?

bam80 replied:
I'm not sure but it might be needed to be able to flash from stock firmware.
I also saw this on other ports to master.

adschm replied:
Yes, but in the other ports it's for sysupgrade from older versions of OpenWrt.

Is the OEM firmware OpenWrt?

bam80 replied:
Hmm, good question.
Judging from https://github.com/kroks-free/openwrt, it mostly is. They have some proprietary modules on top, but overall it looks and feels as OpenWrt clone.
I'll try to upgrade with sysupgrade and we'll see.

bam80 replied:
Sysupgrade went well via vendor's WEB UI, so this line is probably needed so the image could be accepted.

adschm reviewed (changes requested):
This lacks a proper commit message with specifications, flashing instructions and MAC address assignment overview.

I also don't see any board.d files and without the above I cannot tell whether we would need them.

bam80 commented:
> This lacks a proper commit message with specifications, flashing instructions and MAC address assignment overview.

I'm not sure about MAC part, what is the procedure?

bam80 commented:
> I also don't see any board.d files and without the above I cannot tell whether we would need them.

There is a [target/linux/ramips/mt76x8/base-files/etc/board.d/02_network](https://github.com/openwrt/openwrt/pull/4791/files#diff-81a8dc5137c6f1e59f2390960d8eed3fe5d5e2c019e94daff124e70674d09089)

adschm commented:
> > This lacks a proper commit message with specifications, flashing instructions and MAC address assignment overview.
> 
> I'm not sure about the MAC part, what is the procedure?

https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

> > I also don't see any board.d files and without the above I cannot tell whether we would need them.
> 
> There is a [target/linux/ramips/mt76x8/base-files/etc/board.d/02_network](https://github.com/openwrt/openwrt/pull/4791/files#diff-81a8dc5137c6f1e59f2390960d8eed3fe5d5e2c019e94daff124e70674d09089)

Oh, did I overlook it?

bam80 commented on target/linux/ramips/dts/mt7628an_kroks_kndrt31r19.dts:
```diff
@@ -0,0 +1,169 @@
+	#size-cells = <1>;
+
+	macaddr_factory_4: macaddr@4 {
+		reg = <0x4 0x6>;
```
As there is seems only one MAC defined in the DTS, does it mean the others are calculated from this one, or assigned randomly?
Need to check when I get access to the device.

bam80 replied:
Added MAC info in commit message

bam80 commented:
Added info about SIM switching.

The device works normally and fully functional.

bam80 commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -247,6 +247,15 @@ define Device/jotale_js76x8-32m
+  IMAGE_SIZE := 15040k
+  DEVICE_VENDOR := Kroks
+  DEVICE_MODEL := Rt-Pot mXw DS RSIM
+  DEVICE_PACKAGES := kmod-usb2 luci-proto-qmi
```
Well, we could swap it to just `uqmi` but the first time user experience would be a way worse..

bam80 replied:
Do we also need `uboot-envtools` here?

hauke replied:
Please remove the dependency to `luci-proto-qmi` here. It should only depend on packages which are in the main OpenWrt repository `uqmi` should be fine.

bam80 replied:
Done

bam80 replied:
I removed the `luci-proto-qmi` dependency but for some reason still have the QMI interface setting available in LuCI:
![image](https://user-images.githubusercontent.com/233979/185965475-b9bf9526-a6fe-4171-a22b-d1a4761eb573.png)
![image](https://user-images.githubusercontent.com/233979/185965952-42f13880-0135-49d6-a5b8-fb14009d732b.png)
Any idea why?

bam80 replied:
Oh, this qmi proto was cached in browser somehow!
No such thing in Incognito Mode:
![image](https://user-images.githubusercontent.com/233979/185970667-af38484c-fbf3-43c0-8d8c-c16621c8c94a.png)

bam80 committed:
ramips: add support for Kroks Rt-Cse SIM Injector DS

Aka Kroks Rt-Cse5 UW DRSIM (KNdRt31R16), ID 1958:
https://kroks.ru/search/?text=1958
See Kroks OpenWrt fork for support of other models:
https://github.com/kroks-free/openwrt

Device specs:
- CPU: MediaTek MT7628AN
- Flash: 16MB SPI NOR
- RAM: 64MB
- Bootloader: U-Boot
- Ethernet: 5x 10/100 Mbps
- 2.4 GHz: b/g/n SoC
- USB: 1x
- SIM-reader: 2x (driven by a dedicated chip with it's own firmware)
- Buttons: reset
- LEDs: 1x Power, 1x Wi-Fi, 12x others (SIM status, Internet, etc.)

Flashing:
- sysupgrade image via stock firmware WEB interface, IP: 192.168.1.254
- U-Boot launches a WEB server if Reset button is held during power up,
  IP: 192.168.1.1

MAC addresses as verified by OEM firmware:
vendor   OpenWrt   source
LAN      eth0      factory 0x4 (label)
2g       wlan0     label

Signed-off-by: Andrey Butirsky <butirsky@gmail.com>

bam80 commented:
Both devices are tested

bam80 commented:
@adschm ping please?

MaxS0niX commented:
Ping @openwrt, can you please help with the review? What needs to be fixed? What are the further actions?

neheb commented:
Find someone to merge :).

bam80 commented:
@hauke ping, please?

hauke commented:
Please rebase this on top of current master

bam80 commented:
> Please rebase this on top of current master

@hauke done

bam80 committed:
ramips: add support for Kroks Rt-Pot mXw DS RSIM router

Aka "Kroks KNdRt31R19".
Ported from v19.07.8 of OpenWrt fork:
see https://github.com/kroks-free/openwrt
for support of other models.

Device specs:
- CPU: MediaTek MT7628AN
- Flash: 16MB SPI NOR
- RAM: 64MB
- Bootloader: U-Boot
- Ethernet: 1x 10/100 Mbps
- 2.4 GHz: b/g/n SoC
- mPCIe: 1x (usually equipped with an LTE modem by vendor)
- Buttons: reset
- LEDs: 1x Modem, 1x Injector, 1x Wi-Fi, 1x Status

Flashing:
- sysupgrade image via stock firmware WEB interface.
- U-Boot launches a WEB server if Reset button is held during power up.
Server IP: 192.168.1.1

SIM card switching:
The device supports up to 4 SIM cards - 2 locally on board and 2 on
remote SIM-injector.
By default, 1-st local SIM is active.
To switch to e.g. 1-st remote SIM:
echo 0 > /sys/class/gpio/modem1power/value
echo 0 > /sys/class/gpio/modem1sim1/value
echo 1 > /sys/class/gpio/modem1rsim1/value
echo 1 > /sys/class/gpio/modem1power/value

MAC addresses as verified by OEM firmware:
vendor   OpenWrt   source
LAN      eth0      factory 0x4 (label)
2g       wlan0     label

Signed-off-by: Kroks <dev@kroks.ru>
[butirsky@gmail.com: port to master; drop dts-v1]
Signed-off-by: Andrey Butirsky <butirsky@gmail.com>

bam80 commented:
@hauke I addressed the [Test Formalities](https://github.com/openwrt/openwrt/runs/8237331141?check_suite_focus=true#logs) failure, could you re-run the workflow?

hauke commented:
Thank you for your patches, I applied them to master in 0a79c77a4e9d33d352974874a8860c7412e49bd5 and 5806914794b22cd7ff6c47bed9b54e0a5c4234f3.

bam80 commented:
@hauke many thanks for making it happen and for review!
